metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -1708,6 +1708,13 @@ class MetalsLanguageServer(
     }
   }
 
+  @JsonRequest("metals/buildTargets")
```

Comment:
Is there any other use case than specifying it when running/debugging? From what I've seen idea of the `target` actually confuses users and besides it would be needed in some very niche use cases.

Also, it might be easier to name it `listBuildTargets` and leave `buildTagets` name for the current purpose.

Reply:
Sure switching it to `listBuildTargets` shouldn't be an issue. Plus, related to the first point, this isn't really something that would be directly shown to a user without a bit more context to ensure it's clear.

> Is there any other use case than specifying it when running/debugging?

Not really any _concrete_ things at the moment, but I'm playing around with a few things related to displaying a build to a user in a way that is agnostic to the build tool. Mainly I just think there is a ton of information contained in `buildTargets` about your build. All of this _is_ available through your build tool, but I think it's quite powerful to have a way to access this is a way that may be easier than through the build tool, and also agnostic to the actual build tool. I like the idea that if a user is curious about what compiler options they have whether they are using mill or sbt, it doesn't matter, we have that information in a that could just be returned to a client, and a client could show it the same regardless of the build tool.

We don't really have any endpoints like this atm that just give the client a bunch of info that they can start to play around with and come up with useful ways to either display it or help the user with their build. _Maybe_ the closes thing would be `workspace/sybmol` where a huge list of symbols can just be returned to the client to do whatever it wants with.

More so than any use case I currently have for this, I more curious about just exposing this and seeing what comes of it. At worse, it's just not used. At best, we get some cool client features that can maybe do think we didn't really think about before.

Reply:
Also for a bit more context, one thing I was thinking about was right now when you send in a `debug-adapter-start` a main class is required. While that is somewhat easy most of the time to find, it also may not be obvious especially in projects that have multiple modules. So the flow that I was thinking of is that in a situation with no code lenses a user could do maybe a `:MetalsFindMainsMethods` or something like that. Under the hood that would do a `metals/listBuildTargets` to which they could choose `app` or `foo` or whatever they want, then that would send in a `metals/findMainMethodsForTarget` or something like that, which would then return the main methods for that target. So in theory a user could go from opening a project to getting the main method to running right away without having to manually do anything.

However, _maybe_ an easier way to accomplish that is to allow for `debug-adapter-start` to also only allow for `buildTarget` to be set, and then do a look up for the main method in that target.

Reply:
I think allowing to have just a build target set would make it more general across the clients. We could even allow have nothing specified at all and use quick pick to select the main class to run. 

There is a lot of possibilities really 😅 and we could hugely improve the user experience here. 

We do actually also expose build target information via the tree view. We could add a lot more information there like scala options etc.

Reply:
> We could even allow have nothing specified at all and use quick pick to select the main class to run.

Yea this is an interesting idea. I've been looking a lot at the jdtls extensions, and `nvim-jdtls` utilizes the `vscode.java.resolveMainClass` when they setup DAP and it just returns the main classes for this to do what they want with, which is nice.

https://github.com/mfussenegger/nvim-jdtls/blob/a35ce80f0e618da0eceb0d5f6874608d58d3ec26/lua/jdtls.lua#L983

> We do actually also expose build target information via the tree view. We could add a lot more information there like scala options etc.

Indeed we do, but this also forces a user to access it through the tree view, which a certain client may not have.

Anyways since I'm working on DAP stuff currently I may try to play around with this a bit and just try to provide an option that if `debug-adapter-start` is triggered with nothing, we give them a quick pick like said. From poking around, I don't think that would actually be that hard.